frame/staking/src/migrations.rs

```diff
@@ -18,7 +18,87 @@
 
 use super::*;
 use frame_election_provider_support::SortedListProvider;
-use frame_support::traits::OnRuntimeUpgrade;
+use frame_support::{
+	dispatch::GetStorageVersion, pallet_prelude::ValueQuery, storage_alias,
+	traits::OnRuntimeUpgrade,
+};
+
+/// Used for release versioning upto v12.
+///
+/// Obsolete from v13. Keeping around to make encoding/decoding of old migration code easier.
+#[derive(Encode, Decode, Clone, Copy, PartialEq, Eq, RuntimeDebug, TypeInfo, MaxEncodedLen)]
+enum ObsoleteReleases {
+	V1_0_0Ancient,
+	V2_0_0,
+	V3_0_0,
+	V4_0_0,
+	V5_0_0,  // blockable validators.
+	V6_0_0,  // removal of all storage associated with offchain phragmen.
+	V7_0_0,  // keep track of number of nominators / validators in map
+	V8_0_0,  // populate `VoterList`.
+	V9_0_0,  // inject validators into `VoterList` as well.
+	V10_0_0, // remove `EarliestUnappliedSlash`.
+	V11_0_0, // Move pallet storage prefix, e.g. BagsList -> VoterBagsList
+	V12_0_0, // remove `HistoryDepth`.
+}
+
+impl Default for ObsoleteReleases {
+	fn default() -> Self {
+		ObsoleteReleases::V12_0_0
+	}
+}
+
+/// Alias to the old storage item used for release versioning. Obsolete since v13.
+#[storage_alias]
+type StorageVersion<T: Config> = StorageValue<Pallet<T>, ObsoleteReleases, ValueQuery>;
+
+pub mod v13 {
+	use super::*;
+
+	pub struct MigrateToV13<T>(sp_std::marker::PhantomData<T>);
+	impl<T: Config> OnRuntimeUpgrade for MigrateToV13<T> {
+		#[cfg(feature = "try-runtime")]
+		fn pre_upgrade() -> Result<Vec<u8>, &'static str> {
+			frame_support::ensure!(
+				StorageVersion::<T>::get() == ObsoleteReleases::V12_0_0,
+				"Required v12 before upgrading to v13"
+			);
+
+			Ok(Default::default())
+		}
+
+		fn on_runtime_upgrade() -> Weight {
+			let current = Pallet::<T>::current_storage_version();
+			let onchain = StorageVersion::<T>::get();
+
+			if current == 13 && onchain == ObsoleteReleases::V12_0_0 {
+				StorageVersion::<T>::kill();
+				current.put::<Pallet<T>>();
+
+				log!(info, "v13 applied successfully");
+				T::DbWeight::get().reads_writes(1, 2)
+			} else {
+				log!(warn, "Skipping v13, should be removed");
```

Comment:
this is for cases where current > 13, right? because we check `StorageVersion::<T>::get() == ObsoleteReleases::V12_0_0` in pre-upgrade

Reply:
Correct 👍